tests/functional/test_install.py

```diff
@@ -1527,3 +1528,99 @@ def test_target_install_ignores_distutils_config_install_prefix(script):
         (target - script.base_path) in result.files_created and
         (prefix - script.base_path) not in result.files_created
     ), str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+@pytest.mark.parametrize('pip_name', [
+    'pip',
+    'pip{}'.format(sys.version_info[0]),
+    'pip{}.{}'.format(*sys.version_info[:2]),
+    'pip.exe',
+    'pip{}.exe'.format(sys.version_info[0]),
+    'pip{}.{}.exe'.format(*sys.version_info[:2])
+])
+def test_protect_pip_from_modification_on_windows(script, pip_name):
+    """
+    Test ``pip install --upgrade pip`` is raised an error on Windows.
+    """
+    command = [pip_name, 'install', '--upgrade', 'pip']
+    result = script.run(*command, expect_error=True)
+    assert result.returncode != 0
+    new_command = [sys.executable, '-m', 'pip'] + command[1:]
+    assert 'To modify pip, please run the following command:\n{}'.format(
+        ' '.join(new_command)) in result.stderr, str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+def test_protect_pip_from_modification_via_deps_on_windows(script, with_wheel):
+    """
+    Test ``pip install pkga`` is raised an error on Windows
+    if `pkga` implicitly tries to upgrade pip.
+    """
+    # Make a wheel for pkga which requires pip
+    script.scratch_path.joinpath('pkga').mkdir()
+    pkga_path = script.scratch_path / 'pkga'
+    pkga_path.joinpath('setup.py').write_text(textwrap.dedent("""
```

Comment:
What do you think about putting wheel creation into a fixture? Maybe something like

```python
@pytest.fixture
def wheel_factory(script, tmpdir_factory, with_wheel):
    def make_wheel(name, version, dependencies):
        d = tmpdir_factory()
        d.joinpath('setup.py').write_text(textwrap.dedent("""
            from setuptools import setup
            setup(name={!r},
                  version={!r},
                  install_requires={!r})
        """).format(name, version, dependencies))
        script.run(
            'python', 'setup.py', 'bdist_wheel', '--universal', cwd=d
        )
        return d / dist / '{}-{}-py2.py3-none-any.whl'.format(name, version)

    return make_wheel
```

Then we can also remove the comment above (`Make a wheel for pkga which requires pip`) since it should be clear from the test itself

```python
pkga_wheel_path = wheel_factory('pkga', '0.1', ['pip<{}'.format(pip.__version__)])
```

Reply:
> ```python
> wheel_factory
> ```

We already have this -- `tests.lib.create_basic_wheel_for_package`, without the overhead of an additional subprocess call.

Reply:
Marking this as resolved since I've also pointed this out, in my review, along with an example of the existing solution.

Reply:
Ah, I should've looked harder, thank you.

Reply:
Hehe. No worries. I only know about it because I wrote it. 🙃